### PR TITLE
Implement MIX message, presence and roster extensions

### PR DESCRIPTION
**Pull request by lnjX**

### CODE REVIEW

File: src/base/QXmppRosterIq.cpp [hunk 1]
```diff
@@ -253,6 +280,34 @@ void QXmppRosterIq::Item::setSubscriptionTypeFromStr(const QString& type)
...
+/// Sets whether this is a MIX channel.
+
+void QXmppRosterIq::Item::setIsMixChannel(bool isMixChannel)
+{
+    m_isMixChannel = isMixChannel;
+}
+
+/// Returns the participant id for this MIX channel.
+
+QString QXmppRosterIq::Item::participantId() const
```
Comment (jlaine) [inline src/base/QXmppRosterIq.cpp]: Should this be `mixParticipantId` to keep everything MIX-related with a consistent prefix?

Comment (lnjX) [reply src/base/QXmppRosterIq.cpp]: done

File: src/base/QXmppRosterIq.h [hunk 2]
```diff
@@ -86,6 +93,9 @@ class QXMPP_EXPORT QXmppRosterIq : public QXmppIq
         // can be subscribe/unsubscribe (attribute "ask")
         QString m_subscriptionStatus;
         QSet<QString> m_groups;
+        // XEP-0405: Mediated Information eXchange (MIX): Participant Server Requirements
+        bool m_isMixChannel = false;
```
Comment (jlaine) [inline src/base/QXmppRosterIq.h]: This breaks ABI compatibility. If we're going to do this let's at least future-proof it by moving everything to a d-pointer.

File: src/base/QXmppRosterIq.h [hunk 3]
```diff
@@ -108,6 +122,8 @@ class QXMPP_EXPORT QXmppRosterIq : public QXmppIq
     QList<Item> m_items;
     // XEP-0237 Roster Versioning
     QString m_version;
+    // XEP-0405: Mediated Information eXchange (MIX): Participant Server Requirements
```
Comment (jlaine) [inline src/base/QXmppRosterIq.h]: Same as above: let's move everything behind a d-pointer.

File: src/base/QXmppRosterIq.h (hunk 2 above)
Comment (tehnick) [reply src/base/QXmppRosterIq.h]: Useful link about d-pointer and about ABI compatibility in common: 
https://community.kde.org/Policies/Binary_Compatibility_Issues_With_C%2B%2B

And I would prefer the separate commits for this task. For not mixing this with changes related to providing of new features in this PR...

File: src/base/QXmppRosterIq.h (hunk 3 above)
Comment (lnjX) [reply src/base/QXmppRosterIq.h]: I can do that after https://github.com/qxmpp-project/qxmpp/pull/189 has been merged.

Comment (lnjX) [reply src/base/QXmppRosterIq.h]: done

File: src/base/QXmppRosterIq.h (hunk 2 above)
Comment (lnjX) [reply src/base/QXmppRosterIq.h]: done

**Comment by lnjX** [comment]
If the ABI break is a problem for you, you could maybe merge everything into a new develop branch or something like this.

**Comment by tehnick** [comment]
I do not think that ABI breaks may be a problem for now. Because we only at the beginning of a way of introducing the practice of maintaining of stable ABI in qxmpp.

**Comment by lnjX** [comment]
Is there anything because this pull request can't be merged?

**Comment by lnjX** [comment]
@jlaine Please come back from your reviewing holidays :)

**Comment by lnjX** [comment]
Updated implementation to latest XEP version.

File: CMakeLists.txt [hunk 4]
```diff
@@ -5,7 +5,7 @@ cmake_minimum_required(VERSION 3.2)
 set(VERSION_MAJOR 1)
 set(VERSION_MINOR 0)
 set(VERSION_PATCH 1)
-set(SO_VERSION 1)
+set(SO_VERSION 2)
```
Comment (jlaine) [inline CMakeLists.txt]: This looks unrelated to this PR

**Comment by jlaine** [comment]
> @jlaine Please come back from your reviewing holidays :)

I wish someone else would take part in the review process.. It's a bit daunting when there are 20 PRs to review!

File: CMakeLists.txt (hunk 4 above)
Comment (lnjX) [reply CMakeLists.txt]: Yes, that's the problem.

**Comment by lnjX** [comment]
@jlaine Thanks for merging! :)

> I wish someone else would take part in the review process.. It's a bit daunting when there are 20 PRs to review!

I'd like to do so, but I also can't review my own pull requests.
